notifier/sendloop.go

```diff
@@ -0,0 +1,245 @@
+		// TODO(queueperam): maybe we should increment dropped?
+		return
+	default:
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	var dropped int
+	// Queue capacity should be significantly larger than a single alert
+	// batch could be.
+	if d := len(alerts) - s.opts.QueueCapacity; d > 0 {
+		s.logger.Warn("Alert batch larger than queue capacity, dropping alerts", "count", d)
+		dropped += d
+		alerts = alerts[d:]
+	}
+
+	// If the queue is full, remove the oldest alerts in favor
+	// of newer ones.
+	if d := (len(s.queue) + len(alerts)) - s.opts.QueueCapacity; d > 0 {
+		s.logger.Warn("Alert notification queue full, dropping alerts", "count", d)
+		dropped += d
+		s.queue = s.queue[d:]
+	}
+
+	s.queue = append(s.queue, alerts...)
+
+	// Notify sending goroutine that there are alerts to be processed.
+	// If we cannot send on the channel, it means the signal already exists
+	// and has not been consumed yet.
+	s.notifyWork()
+
+	s.metrics.queueLength.WithLabelValues(s.alertmanagerURL).Set(float64(len(s.queue)))
+	if dropped > 0 {
+		s.metrics.dropped.WithLabelValues(s.alertmanagerURL).Add(float64(dropped))
+	}
+}
+
+func (s *sendLoop) notifyWork() {
+	select {
+	case <-s.stopped:
+		return
+	case s.hasWork <- struct{}{}:
+	default:
+	}
+}
+
+func (s *sendLoop) stop() {
+	s.stopOnce.Do(func() {
+		close(s.stopped)
+
+		if s.opts.DrainOnShutdown {
+			s.drainQueue()
+		}
+	})
+}
+
+func (s *sendLoop) drainQueue() {
+	for s.queueLen() > 0 {
+		s.sendOneBatch()
+	}
+}
+
+func (s *sendLoop) queueLen() int {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	return len(s.queue)
+}
+
+func (s *sendLoop) nextBatch() []*Alert {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	var alerts []*Alert
+	if maxBatchSize := s.opts.MaxBatchSize; len(s.queue) > maxBatchSize {
+		alerts = append(make([]*Alert, 0, maxBatchSize), s.queue[:maxBatchSize]...)
+		s.queue = s.queue[maxBatchSize:]
+	} else {
+		alerts = append(make([]*Alert, 0, len(s.queue)), s.queue...)
+		s.queue = s.queue[:0]
+	}
+
+	return alerts
+}
+
+func (s *sendLoop) sendOneBatch() {
+	alerts := s.nextBatch()
+
+	if !s.sendAll(alerts) {
+		s.metrics.dropped.WithLabelValues(s.alertmanagerURL).Add(float64(len(alerts)))
+	}
+}
+
+// loop continuously consumes the notifications queue and sends alerts to
+// the Alertmanager.
+func (s *sendLoop) loop() {
+	for {
+		// If we've been asked to stop, that takes priority over sending any further notifications.
+		select {
+		case <-s.stopped:
+			return
+		default:
+			select {
+			case <-s.stopped:
+				return
+			case <-s.hasWork:
+				s.sendOneBatch()
+
+				// If the queue still has items left, kick off the next iteration.
+				if s.queueLen() > 0 {
+					s.notifyWork()
+				}
+			}
+		}
+	}
+}
+
+func (s *sendLoop) sendAll(alerts []*Alert) bool {
+	if len(alerts) == 0 {
+		return true
+	}
+
+	begin := time.Now()
+
+	var payload []byte
+	var err error
+	switch s.cfg.APIVersion {
+	case config.AlertmanagerAPIVersionV2:
+		openAPIAlerts := alertsToOpenAPIAlerts(alerts)
```

Comment:
we'll need to benchmark this `alertsToOpenAPIAlerts+marshal` per sendloop now.
on main, even though the logic was really buggy, the values and computed once and cached (if no relabeling needed)